horovod/common/fusion_buffer_manager.cc

```diff
@@ -0,0 +1,54 @@
+// Copyright 2018 Uber Technologies, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include "fusion_buffer_manager.h"
+
+namespace horovod {
+namespace common {
+
+Status FusionBufferManager::InitializeBuffer(int64_t threshold, int device, std::shared_ptr<OpContext> context,
+                                             std::function<void()> on_start_init,
+                                             std::function<void()> on_end_init) {
+  auto& elem = tensor_fusion_buffers_[std::make_tuple(device, context->framework())];
+  auto& buffer = elem.first;
+  int64_t& size = elem.second;
+  if (size != threshold) {
+    buffer.reset();
```

Comment:
Can you comment what does `buffer.reset()` do precisely?

If this does not set buffer to nullptr, would you re-initialize it with AllocatePersistent?

Reply:
It will (1) delete the currently held pointer, then (2) set that current pointer reference to `nullptr`.  One thing I can do is explicitly reset it to `nullptr` like: `buffer.reset(nullptr)`.  Would that be preferred?

http://www.cplusplus.com/reference/memory/unique_ptr/reset/